File: tunnel-server-go/internal/control/readgate.go
package control

import "sync"

// ReadGate is a single-connection "are reads armed" switch. When paused, the read loop must
// stop pulling frames until Resume is called; this lets a slow downstream apply TCP
// backpressure to the peer without dropping bytes. It mirrors the C# ReadGate.
type ReadGate struct {
	mu     sync.Mutex
	paused bool
	wait   chan struct{}
}

// NewReadGate returns an armed (not paused) gate.
func NewReadGate() *ReadGate {
	return &ReadGate{}
}

// Pause stops reads. Idempotent.
func (g *ReadGate) Pause() {
	g.mu.Lock()
	defer g.mu.Unlock()
	if g.paused {
		return
	}
	g.paused = true
	g.wait = make(chan struct{})
}

// Resume re-arms reads, returning true if this call transitioned paused -> running.
func (g *ReadGate) Resume() bool {
	g.mu.Lock()
	defer g.mu.Unlock()
	if !g.paused {
		return false
	}
	g.paused = false
	if g.wait != nil {
		close(g.wait)
		g.wait = nil
	}
	return true
}

// IsPaused reports the current state.
func (g *ReadGate) IsPaused() bool {
	g.mu.Lock()
	defer g.mu.Unlock()
	return g.paused
}

// Wait blocks while paused, returning when resumed or when done is closed.
func (g *ReadGate) Wait(done <-chan struct{}) {
	g.mu.Lock()
	if !g.paused {
		g.mu.Unlock()
		return
	}
	wait := g.wait
	g.mu.Unlock()
	select {
	case <-wait:
	case <-done:
	}
}

File: implementations/csharp/server/src/Specus.Server/Networking/ReadGate.cs
namespace Specus.Server.Networking;

/// <summary>
/// A single-connection "is the kernel-side read armed" gate. The Java port's equivalent is
/// <c>ChannelBackpressure.setAutoRead</c> — when you flip <c>autoRead</c> off, Netty stops
/// arming the next <c>read</c>, the TCP receive buffer fills up, and the peer observes a
/// clogged socket without dropping any bytes.
///
/// <para>On .NET there is no built-in <c>autoRead</c>, so this class manages the same state
/// explicitly. The contract:
/// <list type="bullet">
/// <item><see cref="Pause"/> flips the gate off; the reader MUST stop arming <c>ReceiveAsync</c>
/// until <see cref="Resume"/> is called.</item>
/// <item><see cref="WaitIfPausedAsync"/> waits on the current pause token. <see cref="Resume"/>
/// flips the gate on and releases every waiter. If resume wins the race before the reader
/// reaches the gate, the reader simply observes the unpaused state and keeps moving.</item>
/// </list></para>
///
/// <para>The gate is intentionally tiny — we want it observable from many call sites (the
/// reader, the dispatcher, the NAT session) without having to thread the state through a
/// channel context.</para>
/// </summary>
public sealed class ReadGate
{
    private readonly CancellationToken _lifetime;
    private readonly object _sync = new();
    private TaskCompletionSource? _resumeSignal;

    /// <summary>
    /// Snapshot of the read state. <c>true</c> means reads are PAUSED — do not arm another
    /// receive until someone resumes. <c>false</c> means armed (default).
    /// </summary>
    private bool _isPaused;

    public ReadGate(CancellationToken lifetime)
    {
        _lifetime = lifetime;
    }

    public bool IsPaused
    {
        get
        {
            lock (_sync)
            {
                return _isPaused;
            }
        }
    }

    /// <summary>Pause reads. Idempotent — calling when already paused is a no-op.</summary>
    public void Pause()
    {
        if (_lifetime.IsCancellationRequested)
        {
            return;
        }

        lock (_sync)
        {
            if (_isPaused)
            {
                return;
            }

            _isPaused = true;
            _resumeSignal = CreateResumeSignal();
        }
    }

    /// <summary>Waits while reads are paused. Returns immediately when the gate is open.</summary>
    public async ValueTask WaitIfPausedAsync(CancellationToken cancellationToken)
    {
        Task? waitTask;
        lock (_sync)
        {
            waitTask = _isPaused ? _resumeSignal?.Task : null;
        }

        if (waitTask is not null)
        {
            await waitTask.WaitAsync(cancellationToken).ConfigureAwait(false);
        }
    }

    /// <summary>
    /// Resume reads. Returns true if this call un-paused the gate; false if it was already
    /// running or the connection has gone down.
    /// </summary>
    public bool Resume()
    {
        TaskCompletionSource? signal;
        lock (_sync)
        {
            if (!_isPaused)
            {
                return false;
            }

            _isPaused = false;
            signal = _resumeSignal;
            _resumeSignal = null;
        }

        signal?.TrySetResult();

        if (_lifetime.IsCancellationRequested)
        {
            return false;
        }

        return true;
    }

    private static TaskCompletionSource CreateResumeSignal() =>
        new(TaskCreationOptions.RunContinuationsAsynchronously);
}
